[GOFF] Emit symbols for functions.

redstar opened the pull request:
A function entry is mapped to a LD symbol with an offset to the begin of the section.

redstar committed:
[GOFF] Emit symbols for functions.

A function entry is mapped to a LD symbol with an offset to the begin of the section.

MaskRay reviewed (commented):
LGTM! But I will delegated to a GOFF expert for approval...

uweigand reviewed (commented):
There doesn't appear to be any asm output for these?

redstar committed:
Add HLASM output and external references

Adds HLASM output and tests for it, per reviewer comment.
Also adds external references, because it fits very well
into the implementation.

redstar committed:
Fix some formatting issues.

redstar committed:
ER symbols also need to set AMODE

redstar committed:
Add linkage attribute

redstar committed:
Update the asm output for the linkage type

uweigand commented on llvm/include/llvm/MC/MCDirectives.h:
```diff
@@ -48,6 +48,10 @@ enum MCSymbolAttr {
   MCSA_WeakAntiDep,             ///< .weak_anti_dep (COFF)
   MCSA_Memtag,                  ///< .memtag (ELF)
+  MCSA_Code,                    ///< symbol is code (GOFF)
+  MCSA_Data,                    ///< symbol is data (GOFF)
```
The above two duplicate the semantics of `MCSA_ELF_TypeFunction` and `MCSA_ELF_TypeObject` in a sense.  I'm wondering if we can hook into that (already existing) common code logic ...

uweigand commented on llvm/lib/Target/SystemZ/MCTargetDesc/SystemZHLASMAsmStreamer.cpp:
```diff
@@ -183,17 +190,138 @@ void SystemZHLASMAsmStreamer::emitInstruction(const MCInst &Inst,
+}
+
+bool SystemZHLASMAsmStreamer::emitSymbolAttribute(MCSymbol *Sym,
+                                                  MCSymbolAttr Attribute) {
```
It is a bit odd the the asm version of `emitSymbolAttribute` doesn't actually emit any assembler, but just does the same thing as the object file version.  That may be unavoidable as HLASM doesn't have distinct directives but you have to group everything into a single `XATTR`.  But then maybe the two streamers should actually share the implementation somehow?

MaskRay replied:
If the AsmStreamer doesn't do anything, you can remove these symbol property adjustment - making this function a no-op.

uweigand commented on llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:
```diff
@@ -1112,6 +1113,23 @@ void SystemZAsmPrinter::emitEndOfAsmFile(Module &M) {
     emitIDRLSection(M);
+    // Emit EXTRN declarations.
+    OutStreamer->pushSection();
+    for (auto &GO : M.global_objects()) {
```
It's a bit unfortunate if we have to have this loop here.  It looks like `AsmPrinter::doFinalization` is already doing very similar things for other platforms, might be nicer if we could hook into that.

redstar committed:
Replace MCSA_Code/Data

Use MCSA_ELF_TypeFunction/TypeObject instead.

redstar committed:
Remove owner from MCSymbolGOFF

It's only needed on the binary object path, pushing the section into the GOFFObjectWriter is enough.

redstar committed:
Remove LDAttr/ERAttr and initAttributes.

redstar committed:
Fix formatting

redstar committed:
Centralize setting of symbol attributes

redstar committed:
Remove unused emitExterns()

redstar committed:
Completely remove attributes from MCSymbolGOFF

redstar commented on llvm/lib/MC/MCSymbolGOFF.cpp:
```diff
@@ -0,0 +1,46 @@
+
+using namespace llvm;
+
+void MCSymbolGOFF::initAttributes() {
```
I removed the function, the information is now taken from the flags and the other fields.

redstar committed:
Remove left-over popSection()

redstar commented on llvm/lib/Target/SystemZ/SystemZAsmPrinter.cpp:
```diff
@@ -1112,6 +1113,23 @@ void SystemZAsmPrinter::emitEndOfAsmFile(Module &M) {
+    }
+    OutStreamer->switchSection(
+        static_cast<MCSectionGOFF *>(getObjFileLowering().getTextSection())
+            ->getParent());
```
The ownership is only needed for the object path. I sank the RootSD into the GOFF writer, which makes the section switch unnecessary.

redstar commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -37,6 +55,82 @@ void MCGOFFStreamer::changeSection(MCSection *Section, uint32_t Subsection) {
+    Symbol->setExternal(false);
+    break;
+  case MCSA_Weak:
+  case MCSA_WeakReference:
```
Tony currently works on `emitGlobalAlias()`, he will also provide a test case.

redstar committed:
Simplify the implementation by moving some code into finishImpl()

redstar committed:
Move loop to AsmPrinter

redstar commented on llvm/test/CodeGen/SystemZ/zos-no-eh-label.ll:
```diff
@@ -1,4 +1,4 @@
-; RUN: llc -mtriple s390x-ibm-zos < %s | FileCheck %s
+; RUN: llc -mtriple s390x-ibm-zos -emit-gnuas-syntax-on-zos=false < %s | FileCheck %s
```
Well, soon. I would like to fix an "inaccuracy" with relocations first.

uweigand commented on llvm/lib/MC/MCGOFFStreamer.cpp:
```diff
@@ -37,6 +117,15 @@ void MCGOFFStreamer::changeSection(MCSection *Section, uint32_t Subsection) {
 
+void MCGOFFStreamer::emitLabel(MCSymbol *Symbol, SMLoc Loc) {
+  MCObjectStreamer::emitLabel(Symbol, Loc);
+}
```
If you just don't override this function, it should have the same effect, right?

redstar replied:
Yeah, leftover from some early code. Removed.

uweigand reviewed (commented):
Thanks, this does look a lot cleaner.  A few additional comment inline.

redstar committed:
Address most review comments

redstar committed:
Remove hasLDAttributes() and hasERAttributes()

Just inline and simplify the expressions.

redstar committed:
Remove MCSA_WeakReference/MCSA_Global from the loop

redstar committed:
Remove loop with type information

redstar commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -2937,6 +2937,21 @@ bool AsmPrinter::doFinalization(Module &M) {
+        MCSymbol *Sym = TM.getSymbol(&GO);
+        OutStreamer->emitSymbolAttribute(Sym, GO.hasExternalWeakLinkage()
+                                                  ? MCSA_WeakReference
+                                                  : MCSA_Global);
```
It seems that in general, most symbol attributes are not emitted when the symbol is a reference to an external entity. As you noted, the exception is `MCSA_WeakReference`. I solved this by setting the `WeakRefDirective`, and checking `isDefined()` to determine if a symbol is an external references.

redstar commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -2937,6 +2937,21 @@ bool AsmPrinter::doFinalization(Module &M) {
+                                                  : MCSA_Global);
+        OutStreamer->emitSymbolAttribute(Sym, isa<Function>(GO)
+                                                  ? MCSA_ELF_TypeFunction
+                                                  : MCSA_ELF_TypeObject);
```
Same problem as above: those type information is not emitted for external references.
However, the binder matches a symbol and a reference when either one of them has no code/data specified, or if both have the same code/data type. Which means that leaving the code/data field as unspecified works. I tested this with Open XL. As result, I just removed the code.

redstar committed:
Fix formatting

uweigand reviewed (commented):
Thanks, looking really good now!   Just a few final comments inline.

redstar committed:
Remove isInSection

redstar commented on llvm/lib/Target/SystemZ/MCTargetDesc/SystemZHLASMAsmStreamer.cpp:
```diff
@@ -183,17 +191,81 @@ void SystemZHLASMAsmStreamer::emitInstruction(const MCInst &Inst,
+}
+
+static bool sameNameAsCSECT(MCSymbolGOFF *Sym) {
+  if (!Sym->isTemporary() && Sym->isDefined() && Sym->isInSection()) {
```
Well, there is no support for absolute symbols in HLASM and GOFF, so getting an assertion is IMHO ok. The more interesting question is if we can generate an error during isel when we seen an absolute symbol.

uweigand replied:
I thought you can define an absolute symbol in HLASM using `EQU` - just like you can with `.set` in GAS?  I agree you won't see absolute symbols just from compiling high-level code, but with (inline) assembler, that should be possible.

redstar replied:
You can define an absolute value with `EQU`, e.g.
```
addr EQU 72
```
but that symbol cannot be exported. This does not work:
```
  ENTRY addr
```
because `ENTRY` requires a relocatable symbol. Without `ENTRY`, the symbol is not exposed. A symbol with an absolute value is therefore more like a `#define` in C.
From my understanding, `.set` in GAS is different in that works with symbols which can also be external.

redstar replied:
Ok, I agree that it feels fragile to omit `isInSection`. I added a `isInEDSection()` function as you suggested, and use it where needed. I also rewrote the condition here a bit, because `sameNameAsCSECT` is only used once, and inlining the code leads to not checking some predicates twice.

redstar committed:
Rewrite if conditions

redstar committed:
Add back isInSection()

Also add isInEDSection() as simple check.

redstar committed:
Fix formatting.

uweigand reviewed (approved):
LGTM now, thanks!
